File: 03_recursion/python/05_binary_search_recursive.py
def binary_search(arr, target):
    if not arr:
        return -1
    elif len(arr) == 1:
        return -1 if arr[0] != target else target
    
    mid = len(array) // 2
    
    if arr[mid] == target:
        return target
    elif arr[mid] > target:
        return binary_search(arr[:mid], target)
    else:
        return binary_search(arr[mid+1:], target)

File: 03_recursion/scala/05_binary_search_recursive.scala
def binarySearch(arr: Array[Int], target: Int): Option[Int] = {
  if(arr.isEmpty) return None
  
  val mid = arr.length / 2
  
  if(arr(mid) == target) Some(mid)
  else if(arr(mid) > target) binarySearch(arr.slice(0, mid), target)
  else {
    val recursiveResponse = binarySearch(arr.slice(mid + 1, arr.length), target)
    recursiveResponse match {
      case Some(idx) => Some(mid + 1 + idx)
      case None => None
    }
  }
}

println(binarySearch(Array(6, 7, 8, 9, 10), 8))
println(binarySearch(Array(6, 7, 8, 9, 10), 6))

